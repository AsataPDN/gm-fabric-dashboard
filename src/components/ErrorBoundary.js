import PropTypes from "prop-types";
import React from "react";

class ErrorBoundary extends React.Component {
  static propTypes = {
    children: PropTypes.oneOfType([
      PropTypes.arrayOf(PropTypes.node),
      PropTypes.node
    ]),
    fallBackUI: PropTypes.element
  };

  constructor(props) {
    super(props);
    this.state = {
      error: false,
      info: ""
    };
  }

  componentDidCatch(error, info) {
    this.setState({ error, info });
  }

  // Check if a fallback UI is provided, if not, render a default error
  fallBackUI = () => {
    return this.props.fallBackUI ? (
      this.props.fallBackUI
    ) : (
      <div className="no-routes-found-error">
        <div className="content">
          <span data-uk-icon="icon: warning; ratio: 1.8" />
          <span>{`Error: ${this.state.info}`}</span>
        </div>
      </div>
    );
  };
  // If there is an error, render the fallback UI, else render children
  render() {
    return this.state.error ? this.fallBackUI() : this.props.children;
  }
}

export default ErrorBoundary;
