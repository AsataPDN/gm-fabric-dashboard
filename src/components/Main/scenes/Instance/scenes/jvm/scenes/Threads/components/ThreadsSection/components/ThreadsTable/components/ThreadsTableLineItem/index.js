import { PropTypes } from "prop-types";
import React, { Component } from "react";

import IndicatorIcon from "./components/IndicatorIcon";

import StackTrace from "components/Main/components/StackTrace";
import TableCol from "components/Main/components/TableCol";
import TableColThread from "components/Main/components/TableColThread";
import TableColDaemon from "components/Main/components/TableColDaemon";
import TableDrawerCollapse from "components/Main/components/TableDrawerCollapse";
import TableRow from "components/Main/components/TableRow";

import Icon from "components/Icon";
import Glyph from "components/Glyphs/";

/**
 * Line Item containing threads data. Intended to be child of TreadsTable
 * @export
 * @class ThreadsTableLineItem
 * @extends {Component}
 */
export default class ThreadsTableLineItem extends Component {
  static propTypes = {
    arrIndex: PropTypes.number,
    daemon: PropTypes.bool,
    id: PropTypes.number,
    name: PropTypes.string,
    priority: PropTypes.number,
    stack: PropTypes.array,
    state: PropTypes.string
  };

  state = {
    isOpen: false // Is the drawer with the stacktrace open or not
  };

  /**
   * Takes a state and returns a corresponding color associated with that state
   * @param {string} state
   * @returns string
   * @memberof ThreadsTableLineItem
   */
  indicatorColor(state) {
    switch (state) {
      case "RUNNABLE":
        return "#0aab2a";
      case "WAITING":
      case "TIMED_WAITING":
        return "#F5A623";
      case "TERMINATED":
      case "BLOCKED":
      case "NEW":
        return "#BC1C1C";
      default:
        return "blue";
    }
  }

  /**
   * Toggles the stacktrace drawer open or closed
   * @memberof ThreadsTableLineItem
   */
  toggleStack = () => {
    this.setState({ isOpen: !this.state.isOpen });
  };

  render() {
    const { daemon, id, name, priority, stack, state } = this.props;
    const indicatorIcon = this.indicatorColor(state);

    return (
      <TableRow
        selectable={this.state.isOpen}
        key={id}
        onClick={stack.length > 0 ? this.toggleStack : () => {}}
        onKeyDown={evt => {
          if (stack.length && evt.keyCode === 13) {
            evt.preventDefault();
            this.toggleStack();
          }
        }}
        role="link"
        style={stack.length ? { cursor: "pointer" } : {}}
        tabIndex={0}
      >
        <TableColThread paddingLeft>{`${Number(id)}`}</TableColThread>
        <TableColThread style={{ textAlign: "right" }}>
          <IndicatorIcon alt={state} color={indicatorIcon} diameter={15} />
        </TableColThread>
        <TableColThread>
          {stack.length ? (
            <Icon>
              <Glyph name="Threads" />
            </Icon>
          ) : (
            ""
          )}
        </TableColThread>
        <TableCol style={{ flex: "1 1 auto" }}>{name}</TableCol>
        <TableColDaemon>{daemon ? "Yes" : "No"}</TableColDaemon>
        <TableColDaemon>{priority}</TableColDaemon>
        <TableDrawerCollapse
          isOpened={this.state.isOpen}
          onClick={evt => {
            evt.stopPropagation();
          }}
        >
          <StackTrace>
            <div>{`java.lang.Thread.State: ${state}`}</div>
            {stack.map((value, index) => (
              <div key={index}>{`at ${value}`}</div>
            ))}
          </StackTrace>
        </TableDrawerCollapse>
      </TableRow>
    );
  }
}
