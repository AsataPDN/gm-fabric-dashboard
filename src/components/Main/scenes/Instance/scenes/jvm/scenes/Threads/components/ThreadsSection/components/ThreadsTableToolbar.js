import { Actions } from "jumpstate";
import { PropTypes } from "prop-types";
import React, { Component } from "react";

// Importing external deps from src as WebPack Modules directory
import ButtonGroup from "components/ButtonGroup";
import Button from "components/Button";
import Toolbar from "components/Main/components/TableToolbar/components/Toolbar";
import ToolbarCenter from "components/Main/components/TableToolbar/components/ToolbarCenter";

/**
 * Toolbar containing buttons to control what is rendered in the threads table
 * @class ThreadsTableToolbar
 * @extends {Component}
 */
class ThreadsTableToolbar extends Component {
  static propTypes = {
    threadCounts: PropTypes.object,
    threadsFilter: PropTypes.string
  };
  render() {
    const { threadCounts, threadsFilter } = this.props;
    //TODO: figure out why there's an empty className on the parent DIV
    return (
      <Toolbar>
        <ToolbarCenter>
          <ButtonGroup>
            <Button
              active={threadsFilter === "all"}
              clickAction={() => Actions.setThreadsFilter("all")}
              disabled={!threadCounts.all}
              tabIndex={0}
              label="All Threads"
              suffix={threadCounts.all}
            />

            <Button
              active={threadsFilter === "active"}
              clickAction={() => Actions.setThreadsFilter("active")}
              disabled={!threadCounts.active}
              tabIndex={0}
              label="Active"
              suffix={threadCounts.active}
            />
            <Button
              active={threadsFilter === "idle"}
              clickAction={() => Actions.setThreadsFilter("idle")}
              disabled={!threadCounts.idle}
              tabIndex={0}
              label="Idle"
              suffix={threadCounts.idle}
            />
            <Button
              active={threadsFilter === "stopped"}
              clickAction={() => Actions.setThreadsFilter("stopped")}
              disabled={!threadCounts.stopped}
              tabIndex={0}
              label="Stopped"
              suffix={threadCounts.stopped}
            />
          </ButtonGroup>
        </ToolbarCenter>
      </Toolbar>
    );
  }
}

export default ThreadsTableToolbar;
