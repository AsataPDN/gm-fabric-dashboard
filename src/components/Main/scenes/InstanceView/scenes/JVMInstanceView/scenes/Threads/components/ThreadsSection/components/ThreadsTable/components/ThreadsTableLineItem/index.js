import { PropTypes } from "prop-types";
import React, { Component } from "react";

import StackTrace from "components/Main/components/StackTrace";
import TableCol from "components/Main/components/TableCol";
import TableColThread from "components/Main/components/TableColThread";
import TableColDaemon from "components/Main/components/TableColDaemon";
import TableDrawerCollapse from "components/Main/components/TableDrawerCollapse";
import TableRow from "components/Main/components/TableRow";

import StatusIcon from "components/StatusIcon";
import Icon from "components/Icon";
import Glyph from "components/Glyphs/";

/**
 * Line Item containing threads data. Intended to be child of TreadsTable
 * @export
 * @class ThreadsTableLineItem
 * @extends {Component}
 */
export default class ThreadsTableLineItem extends Component {
  static propTypes = {
    arrIndex: PropTypes.number,
    daemon: PropTypes.bool,
    id: PropTypes.number,
    name: PropTypes.string,
    priority: PropTypes.number,
    stack: PropTypes.array,
    state: PropTypes.string
  };

  state = {
    isOpen: false // Is the drawer with the stacktrace open or not
  };

  /**
   * Takes a state and returns a corresponding status string associated with that state
   * @param {string} state
   * @returns string
   * @memberof ThreadsTableLineItem
   */
  getStatus(state) {
    switch (state) {
      case "RUNNABLE":
        return "stable";
      case "WAITING":
      case "TIMED_WAITING":
        return "warning";
      case "TERMINATED":
      case "BLOCKED":
      case "NEW":
        return "down";
      default:
        return "default";
    }
  }

  /**
   * Toggles the stacktrace drawer open or closed
   * @memberof ThreadsTableLineItem
   */
  toggleStack = () => {
    this.setState({ isOpen: !this.state.isOpen });
  };

  render() {
    const { daemon, id, name, priority, stack, state } = this.props;

    const status = this.getStatus(state);

    return (
      <TableRow
        selectable={this.state.isOpen}
        key={id}
        onClick={stack.length > 0 ? this.toggleStack : () => {}}
        onKeyDown={evt => {
          if (stack.length && evt.keyCode === 13) {
            evt.preventDefault();
            this.toggleStack();
          }
        }}
        role="link"
        style={stack.length ? { cursor: "pointer" } : {}}
        tabIndex={0}
      >
        <TableColThread paddingLeft>{`${Number(id)}`}</TableColThread>
        <TableColThread style={{ textAlign: "center" }}>
          <StatusIcon status={status} />
        </TableColThread>
        <TableColThread style={{ textAlign: "center" }}>
          {stack.length ? (
            <Icon>
              <Glyph name="Threads" />
            </Icon>
          ) : (
            ""
          )}
        </TableColThread>
        <TableCol style={{ flex: "1 1 auto" }}>{name}</TableCol>
        <TableColDaemon>{daemon ? "Yes" : "No"}</TableColDaemon>
        <TableColDaemon>{priority}</TableColDaemon>
        <TableDrawerCollapse
          isOpened={this.state.isOpen}
          onClick={evt => {
            evt.stopPropagation();
          }}
        >
          <StackTrace>
            <div>{`java.lang.Thread.State: ${state}`}</div>
            {stack.map((value, index) => (
              <div key={index}>{`at ${value}`}</div>
            ))}
          </StackTrace>
        </TableDrawerCollapse>
      </TableRow>
    );
  }
}
