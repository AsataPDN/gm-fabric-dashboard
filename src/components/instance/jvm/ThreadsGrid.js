import { Actions } from "jumpstate";
import { PropTypes } from "prop-types";
import React, { Component } from "react";
import { connect } from "react-redux";

import ThreadsSection from "./ThreadsSection";

/**
 * Very redundant feeling wrapper container that contains a ThreadsSection
 * TODO: Refactor to combine section and grid and break out threads control like RoutesTableToolbar
 * @class ThreadsGrid
 * @extends {Component}
 */
class ThreadsGrid extends Component {
  static propTypes = {
    threadsEndpoint: PropTypes.string,
    threadsTable: PropTypes.array
  };

  componentDidMount() {
    // Refresh Threads on every mount
    if (this.props.threadsEndpoint) {
      Actions.fetchThreads(this.props.threadsEndpoint);
    } else {
      Actions.fetchThreads();
    }
  }

  render() {
    const { threadsTable } = this.props;
    return (
      <div className="thread-table-container">
        <ThreadsSection threadsTable={threadsTable} />
      </div>
    );
  }
}

function mapStateToProps({
  metrics: { threadsTable },
  settings: {
    fabricServer,
    selectedService,
    selectedServiceVersion,
    selectedInstance
  }
}) {
  return {
    threadsTable,
    threadsEndpoint:
      fabricServer &&
      selectedService &&
      selectedServiceVersion &&
      selectedInstance
        ? `${fabricServer}/threads/${selectedService}/${selectedServiceVersion}/${selectedInstance}`
        : null
  };
}

export default connect(mapStateToProps)(ThreadsGrid);
