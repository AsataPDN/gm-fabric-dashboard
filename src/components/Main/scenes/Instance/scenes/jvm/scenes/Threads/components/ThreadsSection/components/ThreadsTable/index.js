import { PropTypes } from "prop-types";
import React from "react";

import ThreadsTableLineItem from "./components/ThreadsTableLineItem";

import Table from "components/Main/components/Table";
import TableHeader from "components/Main/components/TableHeader";
import TableBody from "components/Main/components/TableBody";

import TableColHeaderThread from "components/Main/components/TableColHeaderThread";
import TableColHeader from "components/Main/components/TableColHeader";

import TableColDaemon from "components/Main/components/TableColDaemon";

ThreadsTable.propTypes = {
  filteredThreadData: PropTypes.array
};

/**
 * Table of threads containing a header and 0..n lineitems
 * @export
 * @param {any} { filteredThreadData = [] }
 * @returns JSX.Element
 */
export default function ThreadsTable({ filteredThreadData = [] }) {
  return (
    <Table>
      <TableHeader>
        <TableColHeaderThread paddingLeft>ID</TableColHeaderThread>
        <TableColHeaderThread style={{ textAlign: "center" }}>
          State
        </TableColHeaderThread>
        <TableColHeaderThread>Trace</TableColHeaderThread>
        <TableColHeader style={{ flex: "1 1 auto" }}>Name</TableColHeader>
        <TableColDaemon header>Daemon</TableColDaemon>
        <TableColDaemon header>Priority</TableColDaemon>
      </TableHeader>
      <TableBody>
        {filteredThreadData.map(
          ({ daemon, id, name, priority, stack, state }, arrIndex) => {
            return (
              <ThreadsTableLineItem
                {...{ daemon, name, priority, stack, state }}
                arrIndex={arrIndex}
                id={Number(id)}
                key={id}
              />
            );
          }
        )}
      </TableBody>
    </Table>
  );
}
