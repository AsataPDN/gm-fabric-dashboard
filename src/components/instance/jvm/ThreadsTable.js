import { PropTypes } from "prop-types";
import React from "react";

import ThreadsTableLineItem from "./ThreadsTableLineItem";

ThreadsTable.propTypes = {
  filteredThreadData: PropTypes.array
};

/**
 * Table of threads containing a header and 0..n lineitems
 * @export
 * @param {any} { filteredThreadData = [] } 
 * @returns JSX.Element
 */
export default function ThreadsTable({ filteredThreadData = [] }) {
  return (
    <div className="div-table thread-table">
      <div className="div-table-header thread-table-header">
        <div className="thread-table-id">ID</div>
        <div className="thread-table-state">State</div>
        <div className="thread-table-stacktrace">Trace</div>
        <div className="thread-table-name">Name</div>
        <div className="thread-table-daemon">Daemon</div>
        <div className="thread-table-priority">Priority</div>
      </div>
      <ol className="div-table-body thread-table-body">
        {filteredThreadData.map(
          ({ daemon, id, name, priority, stack, state }, arrIndex) => {
            return (
              <ThreadsTableLineItem
                {...{ daemon, name, priority, stack, state }}
                arrIndex={arrIndex}
                id={Number(id)}
                key={id}
              />
            );
          }
        )}
      </ol>
    </div>
  );
}
