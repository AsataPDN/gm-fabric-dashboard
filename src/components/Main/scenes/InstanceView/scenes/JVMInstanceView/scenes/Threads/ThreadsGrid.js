import { Actions } from "jumpstate";
import { PropTypes } from "prop-types";
import React, { Component } from "react";
import { connect } from "react-redux";
import _ from "lodash";
import qs from "query-string";

import ThreadsTable from "./components/ThreadsTable";
import TableToolbar from "components/Main/components/TableToolbar";
import ErrorBoundary from "components/ErrorBoundary";

// Importing external deps from src as WebPack Modules directory
import { getVisibleThreads } from "utils/jvm/selectors";
import {
  routerHistoryShape,
  routerLocationShape,
  routerMatchShape
} from "components/PropTypes";

/**
 * Parent container of ThreadsTable and TableToolbar
 * @class ThreadsGrid
 * @extends {Component}
 */
class ThreadsGrid extends Component {
  static propTypes = {
    fabricServer: PropTypes.string,
    history: routerHistoryShape,
    location: routerLocationShape,
    match: routerMatchShape,
    selectedInstance: PropTypes.string,
    selectedService: PropTypes.string,
    selectedServiceVersion: PropTypes.string,
    threads: PropTypes.array,
    threadsEndpoint: PropTypes.string
  };

  // Options for sort dropdown rendered in TableToolbar
  static sortByOptions = [
    {
      value: "state",
      label: "State"
    },
    {
      value: "name",
      label: "Name"
    },
    {
      value: "id",
      label: "ID"
    }
  ];

  static groupByOptions = [
    {
      value: "state",
      label: "State"
    },
    {
      value: "none",
      label: "None"
    }
  ];

  constructor(props) {
    super(props);
    this.state = {
      filterString: "",
      lastPushedFilterString: "",
      groupByAttribute: "none",
      sortByAttribute: "id",
      ascending: true
    };
    // Debounce
    this.debouncedPushHistory = _.debounce(this._pushHistory, 500);
  }

  componentWillMount() {
    this.popAndDecodeHistory(this.props.location.search);
  }

  componentDidMount() {
    const {
      fabricServer,
      selectedService,
      selectedServiceVersion,
      selectedInstance,
      threadsEndpoint
    } = this.props;

    // If fabricServer is truthy, we are running with a "Fabric Server" discovery service,
    // so we need to dynamically build the endpoint for the threads API. If falsy, we can just
    // pull the static configuration value from Redux via threadsEndpoint.
    // Technically Actions.fetchThreads defaults to the static threadsEndpoint, so this is
    // redundant, but we are showing it here just to be more explicit about the logic.

    if (this.props.fabricServer) {
      Actions.fetchAndStoreInstanceThreads(
        `${fabricServer}/threads/${selectedService}/${selectedServiceVersion}/${
          selectedInstance
        }`
      );
    } else {
      Actions.fetchAndStoreInstanceThreads(threadsEndpoint);
    }
  }

  componentWillReceiveProps(nextProps) {
    // We need to check to see if the query string props are the result of user interaction
    // with the search box. We do that by keeping track of the last thing the search box
    // pushed to the query string and filtering out those props. The only expection to this
    // is which the app router action was POP, which means the user hit the back button or
    // otherwise navigated using the client-side router history
    if (
      nextProps.location.search !== `?${this.state.lastPushedQueryString}` ||
      nextProps.history.action === "POP"
    ) {
      this.popAndDecodeHistory(nextProps.location.search);
    }
  }

  setFilterString = filterString =>
    this.setState({ filterString }, () => {
      this.encodeAndPushHistory();
    });

  setSortByAttribute = sortByAttribute => {
    if (sortByAttribute === this.state.sortByAttribute) {
      this.setState({ ascending: !this.state.ascending }, () => {
        this.encodeAndPushHistory();
      });
    } else {
      this.setState({ sortByAttribute }, () => {
        this.encodeAndPushHistory();
      });
    }
  };

  setGroupByAttribute = groupByAttribute =>
    this.setState({ groupByAttribute }, () => {
      this.encodeAndPushHistory();
    });

  /**
   * encodeAndPushHistory encodes local state as a query string and invokes the debounced
   * version of _pushHistory to periodically write to browser history.
   *
   * @memberof ThreadsGrid
   */
  encodeAndPushHistory = () => {
    // Clean local state
    const filterString = this.state.filterString.trim().toLowerCase();
    const groupBy = this.state.groupByAttribute;
    const sortBy = this.state.sortByAttribute;
    // Only encode the truthy pieces of local state into a form ready to be pushed to the
    // browser's query string. If no local state is truthy, call debouncedPushHistory without
    // an argument to remove the search query from the URL.
    let objToEncode = {};

    if (filterString) {
      objToEncode.filterString = filterString;
    }

    // If sortBy or groupBy are set to anything but the defaults,
    // then push to the query string
    if (sortBy !== "id") {
      objToEncode.sortBy = sortBy;
    }
    if (groupBy !== "none") {
      objToEncode.groupBy = groupBy;
    }

    this.debouncedPushHistory(qs.stringify(objToEncode));
  };

  /**
   * _pushHistory is used to push local state to the browser's query string. This function is not
   * called directly but via encodeAndPushHistory, which uses lodash's debounce to prevent individual
   * key strokes from polluting the browser history API.
   * @memberof ThreadsGrid
   */
  _pushHistory = filterString => {
    // Save a query string to local state as lastPushedFilterString to prevent
    // accidental overwriting of user entry and then push the query string
    // to the browser history
    this.setState({ lastPushedFilterString: filterString }, () => {
      this.props.history.push({
        pathname: this.props.match.url,
        search: filterString
      });
    });
  };

  /**
   * popAndDecodeHistory is used to decode and pull local state from the browser's query string
   * @memberof ThreadsGrid
   */
  popAndDecodeHistory = queryString => {
    // Parse the query string for the filterString parameter
    const { filterString = "", groupBy = "none", sortBy = "id" } = qs.parse(
      queryString
    );

    this.setState({
      filterString,
      groupByAttribute: groupBy,
      sortByAttribute: sortBy
    });
  };

  /**
   * Helper function that takes the threads passed as props
   * and sorts according to how sortByAttribute and ascending
   * are set in the local state object.
   * @param {Array} threads
   * @memberof ThreadsGrid
   */
  sort = threads => {
    const { sortByAttribute, ascending } = this.state;
    const sortOrder = ascending ? ["asc"] : ["desc"];
    // thread["id"] is a string, so we need to convert to an int to sort properly
    const sortFunc = thread => {
      return sortByAttribute === "id"
        ? parseInt(thread["id"], 10)
        : thread[sortByAttribute].toLowerCase();
    };
    return _.orderBy(threads, sortFunc, sortOrder);
  };

  render() {
    const { threads } = this.props;
    const filteredThreads = threads.filter(
      thread =>
        thread.name
          .toLowerCase()
          .indexOf(this.state.filterString.trim().toLowerCase()) !== -1
    );

    return (
      <div>
        <ErrorBoundary>
          <TableToolbar
            searchInputProps={{
              filterString: this.state.filterString,
              setFilterString: this.setFilterString,
              searchPlaceholder: "Search Threads"
            }}
            sortByProps={{
              sortByOptions: ThreadsGrid.sortByOptions,
              sortByAttribute: this.state.sortByAttribute,
              setSortByAttribute: this.setSortByAttribute
            }}
            groupByProps={{
              groupByOptions: ThreadsGrid.groupByOptions,
              groupByAttribute: this.state.groupByAttribute,
              setGroupByAttribute: this.setGroupByAttribute
            }}
          />
          <ThreadsTable
            groupByAttribute={this.state.groupByAttribute}
            filteredThreadData={this.sort(filteredThreads)}
          />
        </ErrorBoundary>
      </div>
    );
  }
}

function mapStateToProps(state) {
  return {
    fabricServer: state.settings.fabricServer,
    threads: getVisibleThreads(state),
    threadsEndpoint: state.settings.threadsEndpoint,
    selectedService: state.fabric.selectedService,
    selectedServiceVersion: state.fabric.selectedServiceVersion,
    selectedInstance: state.fabric.selectedInstance
  };
}

export default connect(mapStateToProps)(ThreadsGrid);
