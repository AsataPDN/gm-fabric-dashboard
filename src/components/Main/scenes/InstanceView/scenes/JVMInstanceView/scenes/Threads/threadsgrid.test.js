import React from "react";
import { mount } from "enzyme";
import { MemoryRouter, Route } from "react-router";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import renderer from "react-test-renderer";

// Utilities
import mockState from "json/mockReduxState";

// Components
import ThreadsGrid from "./index";
import ThreadsTableToolbar from "./components/ThreadsTableToolbar";
import ThreadsTable from "./components/ThreadsTable";
import ErrorBoundary from "components/ErrorBoundary";
import ThreadsTableLineItem from "./components/ThreadsTable/components/ThreadsTableLineItem";

// import Actions
import "services";

const toolBarProps = {
  threadCounts: { active: 1, all: 6, idle: 2, stopped: 3 },
  threadsFilter: "active"
};

const threadsTableProps = {
  filteredThreadData: [
    {
      daemon: true,
      id: "2",
      name: "Test Runnable",
      priority: 8,
      stack: [
        "java.lang.Object.wait(Native Method)",
        "java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:143)",
        "java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:164)",
        "java.lang.ref.Finalizer$FinalizerThread.run(Finalizer.java:209)"
      ],
      state: "RUNNABLE"
    }
  ]
};
// Create a mock store and initialize with mock data
const store = configureStore()(mockState);

// Wrap ThreadsGrid in Memory Router
const RouterWrap = (
  <Provider store={store}>
    <MemoryRouter>
      <Route render={props => <ThreadsGrid {...props} />} />
    </MemoryRouter>
  </Provider>
);

let wrapper;

describe("ThreadsGrid View", () => {
  beforeEach(() => {
    wrapper = mount(RouterWrap);
  });

  test("matches snapshot", () => {
    const tree = renderer.create(RouterWrap).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("renders correct components", () => {
    expect(wrapper.find(ThreadsTableToolbar)).toHaveLength(1);
    expect(wrapper.find(ThreadsTable)).toHaveLength(1);
    expect(wrapper.find(ErrorBoundary)).toHaveLength(1);
    // nested component
    expect(wrapper.find(ThreadsTableLineItem)).toHaveLength(1);
  });

  // Will be updating tests shortly after new TableToolbar is implemented for threads view
  test("passes the correct props down to ThreadsTableToolbar", () => {
    expect(wrapper.find(ThreadsTableToolbar).props()).toMatchObject(
      toolBarProps
    );
  });

  test("passes the correct props down to ThreadsTable", () => {
    expect(wrapper.find(ThreadsTable).props()).toMatchObject(threadsTableProps);
  });
});
