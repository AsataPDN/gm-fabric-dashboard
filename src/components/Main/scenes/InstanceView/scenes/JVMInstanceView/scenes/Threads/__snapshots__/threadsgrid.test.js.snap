// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThreadsGrid View matches snapshot 1`] = `
<div>
  <div>
    <div
      className="sc-kGXeez bSgYKS"
    >
      <div
        className="sc-kpOJdX YlRrM"
      >
        <div
          className="sc-kgoBCf iOKiQh"
        >
          <button
            className="btn"
            disabled={false}
            onClick={[Function]}
            style={undefined}
            tabIndex={0}
            title="All Threads"
          >
            <span
              className="label"
            >
              
              All Threads
              <span
                className="label-suffix"
              >
                6
              </span>
            </span>
          </button>
          <button
            className="btn active"
            disabled={false}
            onClick={[Function]}
            style={undefined}
            tabIndex={0}
            title="Active"
          >
            <span
              className="label"
            >
              
              Active
              <span
                className="label-suffix"
              >
                1
              </span>
            </span>
          </button>
          <button
            className="btn"
            disabled={false}
            onClick={[Function]}
            style={undefined}
            tabIndex={0}
            title="Idle"
          >
            <span
              className="label"
            >
              
              Idle
              <span
                className="label-suffix"
              >
                2
              </span>
            </span>
          </button>
          <button
            className="btn"
            disabled={false}
            onClick={[Function]}
            style={undefined}
            tabIndex={0}
            title="Stopped"
          >
            <span
              className="label"
            >
              
              Stopped
              <span
                className="label-suffix"
              >
                3
              </span>
            </span>
          </button>
        </div>
      </div>
    </div>
    <div
      className="sc-htoDjs jAhemH"
    >
      <div
        className="sc-dnqmqq gIvpQT"
      >
        <div
          className="sc-gqjmRU fJIGBQ"
        >
          ID
        </div>
        <div
          className="sc-gqjmRU gQSUoL"
          style={
            Object {
              "textAlign": "center",
            }
          }
        >
          State
        </div>
        <div
          className="sc-gqjmRU gQSUoL"
        >
          Trace
        </div>
        <div
          className="sc-gZMcBi hVuIxC"
          style={
            Object {
              "flex": "1 1 auto",
            }
          }
        >
          Name
        </div>
        <div
          className="sc-bxivhb kvyvhu"
        >
          Daemon
        </div>
        <div
          className="sc-bxivhb kvyvhu"
        >
          Priority
        </div>
      </div>
      <ol
        className="sc-iwsKbI hzfbcM"
      >
        <li
          className="sc-EHOje cjsqKx"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="link"
          style={
            Object {
              "cursor": "pointer",
            }
          }
          tabIndex={0}
        >
          <div
            className="sc-htpNat iVAdOU"
          >
            2
          </div>
          <div
            className="sc-htpNat eydWWN"
            style={
              Object {
                "textAlign": "center",
              }
            }
          >
            <svg
              aria-labelledby="ariaLabelledby"
              className="sc-gzVnrw XkNQd"
              version="1.1"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
              xmlnsXlink="http://www.w3.org/1999/xlink"
            >
              <g
                className="iconBackground"
                fill="rgb(10, 171, 42)"
                fillOpacity=".3"
                fillRule="evenodd"
                id="Square"
              >
                <path
                  d="M12,21 C7.02943725,21 3,16.9705627 3,12 C3,7.02943725 7.02943725,3 12,3 C16.9705627,3 21,7.02943725 21,12 C21,16.9705627 16.9705627,21 12,21 Z"
                  id="BackgroundCircleSmall"
                />
              </g>
              <g
                className="sc-bZQynM iePCjc"
                fill="currentColor"
                title=""
              >
                <title>
                  
                </title>
                <path
                  d="M11,9.86851709 L11,14.1314829 L14.1972244,12 L11,9.86851709 Z M16,12 L10,16 L10,8 L16,12 Z"
                  id="GlyphRunningSmall"
                />
              </g>
            </svg>
          </div>
          <div
            className="sc-htpNat eydWWN"
            style={
              Object {
                "textAlign": "center",
              }
            }
          >
            <svg
              aria-labelledby="ariaLabelledby"
              className="sc-gzVnrw eCnGhT"
              version="1.1"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
              xmlnsXlink="http://www.w3.org/1999/xlink"
            >
              <g
                className="sc-bZQynM iePCjc"
                fill="currentColor"
                title=""
              >
                <title>
                  
                </title>
                <g
                  className="glyph"
                  fill="currentColor"
                  transform="translate(0 0) scale(1)"
                >
                  <path
                    d="M7,7 L19,7 L19,9 L7,9 L7,7 Z M6,11 L18,11 L18,13 L6,13 L6,11 Z M5,15 L17,15 L17,17 L5,17 L5,15 Z"
                    id="shape"
                  />
                </g>
              </g>
            </svg>
          </div>
          <div
            className="sc-bwzfXH fLpJPH"
            style={
              Object {
                "flex": "1 1 auto",
              }
            }
          >
            Test Runnable
          </div>
          <div
            className="sc-bxivhb fPNLHk"
          >
            Yes
          </div>
          <div
            className="sc-bxivhb fPNLHk"
          >
            8
          </div>
        </li>
      </ol>
    </div>
  </div>
</div>
`;
